### DISCUSSION TITLE
How to use TinyUSB in a STM32H750xx project using Keil as IDE?

### DISCUSSION THREAD

**Discussion by bkaz4283** [hathach/tinyusb#740] [Q&A] [answered]
Hello,

I'm using Keil IDE-Version: µVision V5.29.0.0

Tool Version Numbers:
Toolchain:        MDK-Lite  Version: 5.29.0.0
Toolchain Path:    C:\Keil_v5\ARM\ARMCLANG\Bin
C Compiler:         ArmClang.exe        V6.13.1
Assembler:          Armasm.exe        V6.13.1
Linker/Locator:     ArmLink.exe        V6.13.1
Library Manager:    ArmAr.exe        V6.13.1
Hex Converter:      FromElf.exe        V6.13.1
CPU DLL:               SARMCM3.DLL          V5.29.0.0
Dialog DLL:         TCM.DLL              V1.36.5.0
Target DLL:             UL2CM3.DLL           V1.163.0.0
Dialog DLL:         TCM.DLL              V1.36.5.0

I'm trying to import the TinyUSB code into this project but I'm running into a problem as the repository seems to only include code for STM32Fxx MCUs.  Specifically, I mean the code in /Src/portable/st/stm32_fsdev where dcd_init, dcd_int_enable, and a bunch of other needed functions are defined in ways specific only to STM32Fxx MCUs.

I read in the getting_started.md that "In addition, MCU driver submodule is also needed to provide low-level MCU peripheral's driver. Luckily, it will be fetched if needed when you run the `make` to build your board."

I wonder if this is the problem I'm having, since I'm not using "make" to compile my code but rather using the Keil IDE.

Can anyone help me understand how to fetch the needed modules for the STM32H750xx MCU in a way that I can import them into my project in Keil?

I'm on a PC running Windows 10 BTW.

Thanks!
-Brian

**Comment by hathach** [comment]
run this in root of tinyusb repo

```
git submodule update --init hw/mcu/st
```

**Reply by bkaz4283** [reply]
I forgot to mention, I already did the submodule update and hw/mcu/st is fully populated.  But it doesn't contain any code specific to TinyUSB, it just contains system code that I already had:
hw/mcu/st/cmsis_device_h7/Source/Templates/system_stm32h7xx.c
Am I still missing something?

**Reply by hathach** [reply]
What is your issue again ? Maybe check the stock example makefile could help

**Reply by bkaz4283** [reply]
OK, here is the problem I'm running into specifically.

I added everything I thought I needed to my Keil project and when I try to compile, I'm getting these linker errors:
Error: L6218E: Undefined symbol dcd_edpt_clear_stall (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_edpt_open (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_edpt_stall (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_edpt_xfer (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_init (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_int_disable (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_int_enable (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_remote_wakeup (referred from usbd.o).
Error: L6218E: Undefined symbol dcd_set_address (referred from usbd.o).

So I searched for where these functions are supposed to come from and the only place I can find them is here:
tinyusb/src/portable/st/stm32_fsdev/dcd_stm32_fsdev_pvt_st.h
tinyusb/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c

But these files are specific for STM32Fxx family MCUs and I can't use them in my project.  I was expecting to find source code including these functions which will work with the STM32H750xx family as I wrote in my original post.  The new files that got synced when I updated the submodules for hw/mcu/st do not include any code I don't already have.

I looked at a few Makefiles but it doesn't help as I'm using Keil IDE for my project, not Makefiles.

**Reply by duempel** [reply]
@bkaz4283 do you add the whole TinyUSB source files into your µVision Project or add each file manually? I think you maybe just missed to add the right sources. Your MCU does **not use** the **STM32 FS** USB peripheral. You have to add `tinyusb/src/portable/st/synopsys/dcd_synopsys.c` instead of `dcd_stm32_fsdev*` files. Also make sure you set `CFG_TUSB_MCU` symbol to `OPT_MCU_STM32H7`.

**Reply by bkaz4283** [reply]
@duempel Thank you!  You were right, I was only including the source files I thought I needed.
I never in a million years would have guessed that the proper code resides in that /synopsis directory!  Now I included that file and the project compiles!  Thank you again!

**Comment by hathach** [comment, marked as answer]
@bkaz4283 do you add the whole TinyUSB source files into your µVision Project or add each file manually? I think you maybe just missed to add the right sources. Your MCU does **not use** the **STM32 FS** USB peripheral. You have to add `tinyusb/src/portable/st/synopsys/dcd_synopsys.c` instead of `dcd_stm32_fsdev*` files. Also make sure you set `CFG_TUSB_MCU` symbol to `OPT_MCU_STM32H7`.


_Originally posted by @duempel in https://github.com/hathach/tinyusb/discussions/740#discussioncomment-517141_

**Reply by bkaz4283** [reply]
I already replied to this above in the thread.  I added tinyusb/src/portable/st/synopsys/dcd_synopsys.c to the project 11 days ago and the project compiles finally.  However, I'm running into another problem getting the code to run on the MCU which I'll write about in a separate thread.
